cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java

```diff
@@ -884,6 +885,101 @@ public void deleteFlowQueues(HttpRequest request, HttpResponder responder,
+    }  catch (SecurityException e) {
+      responder.sendStatus(HttpResponseStatus.UNAUTHORIZED);
+    } catch (Throwable e) {
+      LOG.error("Got exception:", e);
+      responder.sendStatus(HttpResponseStatus.INTERNAL_SERVER_ERROR);
+    }
+  }
+
+  @POST
+  @Path("/apps/{app-id}/workflows/{workflow-name}/{run-id}/resume")
+  public void resumeWorkflowRun(HttpRequest request, final HttpResponder responder,
+                                 @PathParam("namespace-id") String namespaceId, @PathParam("app-id") String appId,
+                                 @PathParam("workflow-name") String workflowName, @PathParam("run-id") String runId) {
+
+    try {
+      Id.Program id = Id.Program.from(namespaceId, appId, ProgramType.WORKFLOW, workflowName);
+      ProgramRuntimeService.RuntimeInfo runtimeInfo = findAllRuntimeInfos(namespaceId, appId,
+                                                                          ProgramType.WORKFLOW, workflowName,
+                                                                          runtimeService).get(RunIds.fromString(runId));
+      if (runtimeInfo == null) {
+        sendInvalidResponse(responder, id);
+        return;
+      }
+      AppFabricServiceStatus status = suspendResumeProgram(runtimeInfo.getController(), "resume");
+      responder.sendString(status.getCode(), status.getMessage());
+    }  catch (SecurityException e) {
+      responder.sendStatus(HttpResponseStatus.UNAUTHORIZED);
+    } catch (Throwable e) {
+      LOG.error("Got exception:", e);
+      responder.sendStatus(HttpResponseStatus.INTERNAL_SERVER_ERROR);
+    }
+  }
+
+  private void sendInvalidResponse(HttpResponder responder, Id.Program id) {
+    try {
+      AppFabricServiceStatus status;
+      try {
+        ProgramStatus programStatus = getProgramStatus(id, ProgramType.WORKFLOW);
+        if (programStatus.getStatus().equals(HttpResponseStatus.NOT_FOUND.toString())) {
+          status = AppFabricServiceStatus.PROGRAM_NOT_FOUND;
+        } else if (ProgramController.State.COMPLETED.toString().equals(programStatus.getStatus())
+          || ProgramController.State.KILLED.toString().equals(programStatus.getStatus())
+          || ProgramController.State.ERROR.toString().equals(programStatus.getStatus())) {
+          status = AppFabricServiceStatus.PROGRAM_ALREADY_STOPPED;
+        } else {
+          status = AppFabricServiceStatus.RUNTIME_INFO_NOT_FOUND;
+        }
+      } catch (Exception e) {
+        status = AppFabricServiceStatus.INTERNAL_ERROR;
+      }
+      responder.sendString(status.getCode(), status.getMessage());
+    } catch (SecurityException e) {
+      responder.sendStatus(HttpResponseStatus.UNAUTHORIZED);
+    } catch (Throwable e) {
+      LOG.error("Got exception:", e);
+      responder.sendStatus(HttpResponseStatus.INTERNAL_SERVER_ERROR);
+    }
+  }
+
+  private AppFabricServiceStatus suspendResumeProgram(ProgramController controller, String action) {
```

Comment:
Is it better to split this into ``suspendProgram`` and ``resumeProgram``? I think it may be much cleaner, also callers will not have to pass in exact strings for the particular action. Also seems like we don't have much common logic between suspend and resume.